src/pallene/coder.lua

```diff
@@ -1648,130 +1670,91 @@ end
     else
-        tmpl = [[
-            if ($condition) {
-                ${then_}
-            } else {
-                ${else_}
-            }
-        ]]
+        tagged_union.error(typ._tag)
     end
 
-    return util.render(tmpl, {
-        condition = condition,
-        then_ = then_,
-        else_ = else_,
-    })
-end
-
-gen_cmd["Loop"] = function(self, cmd, func)
-    local body = self:generate_cmd(func, cmd.body)
     return (util.render([[
-        while (1) {
-            ${body}
-        }
+        ${macro}($i, $cond, $iter, $count, $start, $limit, $step)
     ]], {
-        body = body
+        macro = macro,
+        i     = self:c_var(cmd.dst_i),
+        cond  = self:c_var(cmd.dst_cond),
+        iter  = self:c_var(cmd.dst_iter),
+        count = self:c_var(cmd.dst_count),
+        start = self:c_value(cmd.src_start),
+        limit = self:c_value(cmd.src_limit),
+        step  = self:c_value(cmd.src_step),
     }))
 end
 
-gen_cmd["For"] = function(self, cmd, func)
-    local typ = func.vars[cmd.dst].typ
+gen_cmd["IterFor"] = function(self, cmd, func)
+    local typ = func.vars[cmd.dst_i].typ
 
     local macro
     if     typ._tag == "types.T.Integer" then
-        macro = "PALLENE_INT_FOR_LOOP"
+        macro = "PALLENE_INT_ITER_FOR_LOOP"
     elseif typ._tag == "types.T.Float" then
-        macro = "PALLENE_FLT_FOR_LOOP"
+        macro = "PALLENE_FLT_ITER_FOR_LOOP"
     else
         tagged_union.error(typ._tag)
     end
 
     return (util.render([[
-        ${macro}_BEGIN($x, $start, $limit, $step)
-        {
-            $body
-        }
-        ${macro}_END
+        ${macro}($i, $cond, $iter, $count, $start, $limit, $step)
     ]], {
         macro = macro,
-        x     = self:c_var(cmd.dst),
+        i     = self:c_var(cmd.dst_i),
+        cond  = self:c_var(cmd.dst_cond),
+        iter  = self:c_var(cmd.dst_iter),
+        count = self:c_var(cmd.dst_count),
         start = self:c_value(cmd.src_start),
         limit = self:c_value(cmd.src_limit),
         step  = self:c_value(cmd.src_step),
-        body  = self:generate_cmd(func, cmd.body)
     }))
 end
 
+gen_cmd["Jmp"] = function(self, cmd, _func)
+    local jmp = "goto " .. self:c_label(cmd.target) .. ";"
+    return jmp
+end
+
+gen_cmd["JmpIfFalse"] = function(self, cmd, _func)
+    local jmp_cond = util.render("if(!($v)) {goto $l;}", {
+        v = self:c_value(cmd.src_cond),
+        l = self:c_label(cmd.target),
+    })
+    return jmp_cond
+end
+
 gen_cmd["CheckGC"] = function(self, cmd, func)
     return util.render([[ luaC_condGC(L, ${update_stack_top}, (void)0); ]], {
         update_stack_top = self:update_stack_top(func, cmd) })
 end
 
+function Coder:generate_blocks(func)
+    local out = ""
+    for block_id,block in ipairs(func.blocks) do
+        -- putting "(void)0" at the label so the C compiler doesn't complain about dangling labels
+        local content = self:c_label(block_id) .. ":(void)0;\n"
```

Comment:
It looks like we're adding (void)0 to all labels. Is there a way to only add them when necessary?

Reply:
I'll investigate this dangling label error further. Maybe if we only insert it at the last block it will work still.